cmd/juju/cloud/listcredentials.go

```diff
@@ -156,35 +156,64 @@ func (c *listCredentialsCommand) personalClouds() (map[string]jujucloud.Cloud, e
 	return c.personalCloudsFunc()
 }
 
-func (c *listCredentialsCommand) Run(ctxt *cmd.Context) error {
-	var credentials map[string]jujucloud.CloudCredential
-	credentials, err := c.store.AllCredentials()
-	if err != nil && !errors.IsNotFound(err) {
-		return err
-	}
-	if c.cloudName != "" {
-		for cloudName := range credentials {
-			if cloudName != c.cloudName {
-				delete(credentials, cloudName)
-			}
-		}
+func (c *listCredentialsCommand) publicClouds() (map[string]jujucloud.Cloud, error) {
+	publicClouds, _, err := jujucloud.PublicCloudMetadata(jujucloud.JujuPublicCloudsPath())
+	if err != nil {
+		return nil, err
 	}
+	return publicClouds, nil
+}
 
-	// Find local cloud names.
-	personalClouds, err := c.personalClouds()
+func (c *listCredentialsCommand) builtInClouds() (map[string]jujucloud.Cloud, error) {
+	builtin, err := common.BuiltInClouds()
 	if err != nil {
-		return err
+		return nil, errors.Trace(err)
 	}
-	var personalCloudNames []string
-	for name := range personalClouds {
-		personalCloudNames = append(personalCloudNames, name)
+	return builtin, nil
+}
+
+func (c *listCredentialsCommand) Run(ctxt *cmd.Context) error {
+	var cloudNames []string
+
+	if c.cloudName != "" {
+		cloudNames = []string{c.cloudName}
+	} else {
+		// Find cloud names.
```

Comment:
I am not sure how I feel about this whole long stanza here but this area will be changed shortly to also list credential from the controller so maybe it is ok.

Reply:
Yep, I regretted it immediately. My alternate was to alter the CredentialGetter interface to return the names for all the entries, was trying to avoid changing that interface as it then requires the memory store and file store to be updated + unit tests for the additional function. Possibly would be better to add a method to aggregate all the available clouds (personal, public and builtin), as this is almost verbatim to what is in the boostrap --clouds command.